pyproject.toml

```diff
@@ -54,20 +54,36 @@ classifiers=[
 ]
 
 [tool.poetry.dependencies]
-python = ">=3.9,<3.13"
-requests = "^2.31.0"
+python = ">=3.8,<3.13"
 typing-extensions = ">=3.7.4"
-python-dateutil = ">=2.5.3"
-urllib3 = "2.0.7"
+urllib3 = ">=1.26.16"
 tqdm = ">=4.64.1"
 grpcio = { version = ">=1.44.0", optional = true }
 grpc-gateway-protoc-gen-openapiv2 = { version = "0.1.0", optional = true }
 googleapis-common-protos = { version = ">=1.53.0", optional = true }
 lz4 = { version = ">=3.1.3", optional = true }
 protobuf = { version = "~=3.20.0", optional = true }
 
+[tool.poetry.group.types]
+optional = true
+
+[tool.poetry.group.types.dependencies]
+mypy = "^1.6.1"
+types-urllib3 = "^1.26.25.14"
+grpc-stubs = "^1.53.0.3"
+pandas-stubs = [
+    {version = "^2.1.1.230928", python = ">=3.9"},
+    {version = "^1.5.3.230321", python = ">=3.8,<3.9"}
+]
+types-tqdm = "^4.66.0.3"
+types-protobuf = "^4.24.0.4"
```

Comment:
These are only used for running `mypy` type checks. I moved them into a different group hoping that would isolate them from the "prod" dependencies, but the poetry resolver requires all combinations of groups and extras to be compatible so it actually didn't meaningfully help solve my issue. In the end I had to add this additional version config for pandas-stubs in order to bring back python 3.8 support since the 2.1.1.x stubs were python >=3.9 only.

Reply:
Thanks for digging into all this, does not sound fun to have tracked down.